File: appengine/monorail/framework/clientmon.py
# Copyright 2016 The Chromium Authors. All rights reserved.
# Use of this source code is govered by a BSD-style
# license that can be found in the LICENSE file or at
# https://developers.google.com/open-source/licenses/bsd

"""A class to log client-side javascript error reports.

Updates frontend/js_errors ts_mon metric.
"""

import json
import logging

from framework import jsonfeed

from infra_libs import ts_mon

class ClientMonitor(jsonfeed.JsonFeed):
  """JSON feed to track client side js errors in ts_mon."""

  js_errors = ts_mon.CounterMetric('frontend/js_errors',
      'Number of uncaught client-side JS errors.',
      None)

  def HandleRequest(self, mr):
    """Build up a dictionary of data values to use when rendering the page.

    Args:
      mr: commonly used info parsed from the request.

    Returns:
      Dict of values used by EZT for rendering the page.
    """

    post_data = mr.request.POST
    errors = post_data.get('errors')
    try:
      errors = json.loads(errors)

      total_errors = 0
      for error_key in errors:
        total_errors += errors[error_key]
      logging.error('client monitor report (%d): %s', total_errors,
          post_data.get('errors'))
      self.js_errors.increment_by(total_errors)
    except Exception as e:
      logging.error('Problem processing client monitor report: %r', e)

    return {}

File: go/src/infra/appengine/sheriff-o-matic/som/handler/clientmon.go
// Copyright 2016 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package handler

import (
	"encoding/json"
	"net/http"

	"go.chromium.org/luci/common/logging"
	"go.chromium.org/luci/common/tsmon/metric"
	"go.chromium.org/luci/server/auth/xsrf"
	"go.chromium.org/luci/server/router"
)

var (
	jsErrors = metric.NewCounter("frontend/js_errors",
		"Number of uncaught javascript errors.", nil)
)

type eCatcherReq struct {
	Errors    map[string]int64 `json:"errors"`
	XSRFToken string           `json:"xsrf_token"`
}

// PostClientMonHandler handles uncaught javascript errors reported by the client.
func PostClientMonHandler(ctx *router.Context) {
	c, w, r := ctx.Context, ctx.Writer, ctx.Request

	req := &eCatcherReq{}
	if err := json.NewDecoder(r.Body).Decode(req); err != nil {
		errStatus(c, w, http.StatusBadRequest, err.Error())
		return
	}

	if err := xsrf.Check(c, req.XSRFToken); err != nil {
		errStatus(c, w, http.StatusForbidden, err.Error())
		return
	}

	for _, errCount := range req.Errors {
		jsErrors.Add(c, errCount)
	}
	logging.Errorf(c, "clientmon report: %v", req.Errors)
}
